### PR TITLE
OCPCLOUD-2565: Featuregate implementation

### PR DESCRIPTION
**Pull request by theobarberbany**
This change: 

- Updates the MAO, to bring in new changes to the `Machine` Reconciler.
- Reads in feature gates from flags, passing them to the reconciler. 

Running the binary without passing the --feature-gates flag will result in the feature gates initialising as follows: 

```
I0812 17:36:32.555187    1120 main.go:154] Initializing feature gates: AllAlpha=true|false (ALPHA - default=false), AllBeta=true|false (BETA - default=false), MachineAPIMigration=true|false (ALPHA - default=false)
I0812 17:36:32.555275    1120 main.go:159] FeatureGateMachineAPIMigration initialised: false
```

Note that any feature gates that are GA will not be printed in the help text, but can still be set by flags.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -116,3 +118,5 @@ require (
 	sigs.k8s.io/structured-merge-diff/v4 v4.4.1 // indirect
 	sigs.k8s.io/yaml v1.4.0 // indirect
 )
+
+replace github.com/openshift/machine-api-operator => github.com/theobarberbany/machine-api-operator v0.0.0-20240812110752-d170650c72d6
```
Comment (JoelSpeed) [inline go.mod:122]: To be updated?

Comment (theobarberbany) [reply go.mod:122]: Welp, yes! Cheers!

**Comment by JoelSpeed** [comment]
@theobarberbany Why did you add the hold?

**Comment by theobarberbany** [comment]
> @theobarberbany Why did you add the hold?

I've started to look at the aws and azure providers and they both have existing feature gating. Assuming we want all feature gating to be done via flags (I vote yes) I'll have to migrate those, and that'll mean passing them as flags. I don't think there's a nice way from the operator to only pass a feature gate as a flag if it's required by the operand (e.g only pass the Azure Workload Identity flag if it's azure). This means all operands will need to be aware of all feature flags any of them may take. 

I haven't got all of them in the list of known gates for the main cmd here yet so it would mean another PR.

So TL;DR, just to buy myself time to work out all of the feature gates that each provider will need to be aware of!

**Comment by theobarberbany** [comment]
The full set we need is: `apifeatures.FeatureGateVSphereStaticIPs, apifeatures.FeatureGateAzureWorkloadIdentity, apifeatures.FeatureGateGCPLabelsTags, apifeatures.FeatureGateMachineAPIMigration` 

/unhold

**Comment by JoelSpeed** [comment]
@theobarberbany Would it not be better to have it so that the providers just ignore feature gates they aren't aware of? We don't want to have to have GCP feature gates in the AWS codebase ideally

**Comment by theobarberbany** [comment]
@JoelSpeed yep that's my plan - although in order for the cmd to not panic [(unknown feature gate X) ](https://github.com/openshift/machine-api-provider-azure/blob/5d0c8cc99a967bef27a08f8b347071acf40a047c/vendor/k8s.io/component-base/featuregate/feature_gate.go#L231-L233) we need to make the gate aware of the feature :( 

I suppose we could try to parse the error, and ignore if it's an unknown gate, but that may hide legitimate errors (e.g a typo with a gate)

**Comment by JoelSpeed** [comment]
Other components, looking at KCM/KAS specifically, we have patches in so that if they pick up unknown gates, they just ignore them. I assumed we would behave the same here, and not have to worry about switching or selecting a subset of gates at all

**Comment by theobarberbany** [comment]
>  looking at KCM/KAS specifically, we have patches in so that if they pick up unknown gates, they just ignore them.

I'm not sure we do? Or I'm looking in the wrong place 😓 

Using KAS as an example: 

```bash
❯ ./cluster-autoscaler-arm64 -feature-gates=FeatureGateFoo=true
....
unknown shorthand flag: 'f' in -feature-gates=FeatureGateFoo=true
```
I'll take a look at the KCM now

**Comment by theobarberbany** [comment]
PR to extend librarygo to take flags.Flagset https://github.com/openshift/library-go/pull/1775 

Once this merges we can use this :)

**Comment by nrb** [comment]
/retest

CI infra failures caused some of the jobs to fail.

**Comment by nrb** [comment]
/tide refresh

Acknowledge critical fixes should have been lifted.
